docs/design/resources.md

```diff
@@ -0,0 +1,133 @@
+following:
+
+1. If a headless service for the cluster does not exist, create it.
+2. Attempt to dial the cluster using the headless service. If
+   successfully connected update the `status` field with information
+   from the cluster.
+3. If we have communication with the cluster and the cluster is
+   quorate, check that the number of members the cluster expects is
+   the same as the cluster resource's specified `size`. If the cluster
+   has too few members, add a new one using the API. If too many,
+   remove exactly one using the API.
+4. List the `EtcdPeer` resources that are our children. If we have
+   communication with the cluster, reconcile the peer resources with
+   the member list in etcd:
+
+   * If a member exists in the etcd member list, but does not have a
+     corresponding `EtcdPeer`, then create the peer resource.
+   * If a `EtcdPeer` resource exists that does not have a
+     corresponding member in the etcd memeber list, then remove the
+     peer resource.
+
+   If we don't have communication with the cluster (or the cluster is
+   nonquorate) and we don't have enough peer resources compared with
+   our desired size, then add peer resources to make up the expected
+   size.
+5. Update the status resource based on our communication with the
+   cluster. If the cluster cannot be reached we mark ourselves as not
+   ready. If the cluster is not quorate we mark as non quorate.
+
+## Common operations
+
+### Bootstrapping
+
+1. A user creates a `EtcdCluster` resource with size 3.
+2. Using the logic above, the cluster controller creates a headless
+   service, fails to dial the cluster, and creates three peer
+   resources.
+3. Using the logic above, the controller for the peer resource notices
+   that the created peer resources do not have PVCs or pods, and so
+   creates them.
+4. The pods run etcd, bootstrap to each other, and then form a
+   cluster.
+5. The cluster resource eventually is able to communicate with the
+   cluster and comes up.
+
+### Scale-up
+
+1. A user edits the `EtcdCluster` resource to increase the size by
+   one.
+2. Using the logic above, the cluster controller notes that the cluster
+   is undersized by one and adds a new member to the etcd cluster
+   using the etcd API.
+3. The cluster controller then notices that the cluster expects a member
+   that does not have a `EtcdPeer` resource, and so adds one.
+4. The peer controller creates the PVC and pod for the new peer, which
+   then bootstraps and joins the cluster.
+
+### Scale-down
+
+1. A user edits the `EtcdCluster` resource to decrease the size by
+   one.
+2. Using the logic above, the cluster controller notes that the number
+   of members in the etcd cluster is too big by one and removes one.
+3. The cluster controller notices that it has a `EtcdPeer` resource for a
+   peer that the cluster does not know about and deletes one.
+4. A deletion hook for the peer resource removes the pod, but not any
```

Comment:
This is more likely handled by ownerReferences, but again, needn't be changed immediately as we can always come back to this and change it once implemented 😄